Support more test assertions in OptionalEmptinessHandler

XN137 opened the pull request:
supported with JUnit 4:
```
Assert.assertTrue(x.isPresent())
Assert.assertFalse(x.isEmpty())
```

supported with JUnit 5:
```
Assertions.assertTrue(x.isPresent())
Assertions.assertFalse(x.isEmpty())
```

supported with AssertJ:
```
assertThat(x.isPresent()).isTrue()
assertThat(x.isEmpty()).isFalse()

assertThat(x).isPresent()
assertThat(x).isNotEmpty()
```

supported with Truth:
```
assertThat(x.isPresent()).isTrue()
assertThat(x.isEmpty()).isFalse()
```

msridhar commented:
@XN137 thanks for the great contribution!  This one will take a more careful review, and it may take us a few days to get to it.

XN137 commented on nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java:
```diff
@@ -81,9 +84,8 @@ public class OptionalEmptinessHandler extends BaseNoOpHandler {
-        && optionalIsGetCall((Symbol.MethodSymbol) ASTHelpers.getSymbol(expr), state.getTypes())) {
-      return true;
+    if (!exprMayBeNull && expr.getKind() == Tree.Kind.METHOD_INVOCATION) {
+      return optionalIsGetCall((Symbol.MethodSymbol) ASTHelpers.getSymbol(expr), state.getTypes());
```
small refactoring i thought may make sense to safe a few cycles, but happy to revert this part if not worth it (since admittedly hard to review)

XN137 replied:
removed since it may be controversial

XN137 commented on nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java:
```diff
@@ -133,8 +195,9 @@ public NullnessHint onDataflowVisitMethodInvocation(
-        && optionalIsGetCall((Symbol.MethodSymbol) ASTHelpers.getSymbol(expr), state.getTypes())
+    Symbol symbol = ASTHelpers.getSymbol(expr);
+    if (symbol instanceof Symbol.MethodSymbol
+        && optionalIsGetCall((Symbol.MethodSymbol) symbol, state.getTypes())
```
small refactoring to not call `getSymbol` twice

XN137 commented on nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java:
```diff
@@ -168,35 +231,54 @@ public boolean includeApInfoInSavedContext(AccessPath accessPath, VisitorState s
+          && methodNameUtil.isMethodAssertThat(receiverSymbol)) {
+        // assertThat will always have at least one argument, So safe to extract from the arguments
+        Node arg = receiverMethod.getArgument(0);
+        if (arg instanceof LocalVariableNode || arg instanceof FieldAccessNode) {
```
are there any other valid cases we want find here?

i.e. `assertThat(obj.getOptionalField()).isPresent()` ?

XN137 replied:
makes sense, can you provide guidance or existing code that shows how "access path" handling logic can be reused here?
it wasnt immediately clear to me at the `Node instanceof` level :sweat_smile:

XN137 replied:
> (note that the key needs to be a constant for NullAway to understand that multiple calls to .get(...) will retrieve the same value)

wondering: why is that? is that a planned feature to at least support `final String key = "foo"; map.get(key)` or the like?
because that didnt seem to work when i tried writing a test.

msridhar replied:
> > (note that the key needs to be a constant for NullAway to understand that multiple calls to .get(...) will retrieve the same value)
> 
> wondering: why is that? is that a planned feature to at least support `final String key = "foo"; map.get(key)` or the like? because that didnt seem to work when i tried writing a test.

This is not planned currently, but it's mostly a matter of priorities.  If that pattern comes up a lot in your code base (a `final` local variable assigned a string constant), I think we'd be open to a PR.  We'd just have to be careful it doesn't add significantly to NullAway's overhead.

lazaroclapp replied:
Ah, my bad here! We have such a support for `final static` fields and I think an open issue on supporting `final` local vars (see discussion in #697), I just got ahead of myself here 😅

XN137 replied:
thanks for the feedback

>  If that pattern comes up a lot in your code base (a final local variable assigned a string constant), I think we'd be open to a PR

it imo doesnt have to be a string constant, just a type known to be immutable (that cant be re-assigned due to the final keyword).

here is an example of what i was unable to do in the test:
```
final String theKey = "theKey";
Assertions.assertThat(getMapWithOptional().get(theKey)).isNotNull();
Assertions.assertThat(getMapWithOptional().get(theKey)).isPresent();
int y = 1 + getMapWithOptional().get(theKey).get();
```
resulting in:
```
/Test.java:66: warning: [NullAway] Invoking get() on possibly empty Optional getMapWithOptional().get(theKey)
    int y = 1 + getMapWithOptional().get(theKey).get();
                                                ^
```
but it would work when using the same string literal everywhere as key instead.
is the access path not considered equal when a variable is involved like that?
that's what i was wondering about here.

i dont immediately see how this relates to the Stream-issue which adds another layer of complexity.

msridhar replied:
I agree this is unrelated to streams. I don't immediately see any conceptual issues with treating final local variables as constant keys in map access paths. @lazaroclapp do you see an issue? Not too sure about complexity of implementing support or perf cost

lazaroclapp replied:
Ok, on the linked issue, the relevant comment is [this one](https://github.com/uber/NullAway/issues/697#issuecomment-1350129124). But I entirely agree that the context of that issue overcomplicates the discussion. So here is my best attempt of an explanation:

1. During intra-procedural dataflow, there is a set of locations NullAway tracks the nullability of, defined by the concept of an Access Path. This includes: variables (locals, arguments, etc), chains of fields (`this.f.g.h`), chains of method calls that NullAway interprets to be getters through some heuristics (the easies case is zero-argument methods: `this.getF().getG().getH()`), and - as a semi-special case - `Map::get(k)` for a constant key `k`. 
2. Combinations of the above are allowed. A concatenation of an existing valid Access Path followed by one of the above is also a valid Access Path. Here is the [data structure](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java) used, which gives some context about the cases. 
3. What "a constant key `k`" means, though, is maybe a bit hacky. Primitive constants are acknowledged as constant here, as are `static fields` fields. Here is an example [test](https://github.com/uber/NullAway/blob/master/nullaway/src/test/java/com/uber/nullaway/NullAwayAccessPathsTests.java#L81) of what is known to be supported.
4. Now, interesting caveat from your example above. I see that `Assertions.assertThat(getMapWithOptional().get(theKey)).isNotNull()` is being interpreted correctly as guaranteeing the path `getMapWithOptional().get(theKey)` is non-null, or else you should be getting a standard nullness error for that final line, rather than the empty optional issue. No? Or is it just that the optional emptiness error triggers first somehow? If the former, then that is a bit surprising to me, based on what we test for, but to be fair there does seem to be a [more general case](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java#L297) for map keys that would match such expression. 

What I would do, if I were debugging this at this point, would be to check:

1. If the example above, with a `final String thisKey` local variable, works when only checking for the nullability of the map's value, independent of optional emptiness handling.
2. If the example above, with optional emptiness handling on, works when `theKey` is a `static final` field, as in the test linked above.

Either way, I would leave any fix for a follow up PR, because if this ends up involving changing core Access Path definition stuff rather than just this handler, then we need to be very careful about performance/end-to-end testing and it's easier with a smaller change. And, even if it doesn't, it feels like conceptually separate from supporting more test assertion APIs 🙂

XN137 commented on nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java:
```diff
@@ -168,35 +231,54 @@ public boolean includeApInfoInSavedContext(AccessPath accessPath, VisitorState s
+  private MethodInvocationNode maybeUnwrapBooleanValueOf(MethodInvocationNode node) {
+    // Since assertThat(a.isPresent()) changes to
+    // Truth.assertThat(Boolean.valueOf(a.isPresent()))
+    // need to be unwrapped from Boolean.valueOf
```
this looks odd but was already present in the old code and Truth-tests fail without it

XN137 replied:
ah i see thanks, due to the comment i thought this is some Truth specific thing (idk that library), but yeah i it's just auto boxing by the java compiler :+1:

XN137 reviewed (commented):
while trying to introduce nullaway to our codebase i ran into the problem that i found no good way to assert an optional to be present in our test code

XN137 commented:
i will look into the test failures on JDK8 later this week

XN137 commented:
@msridhar `Optional.isEmpty()` only exists on JDK11+, is there mechanism to make this JDK distinction in the tests somehow?

msridhar commented:
> @msridhar `Optional.isEmpty()` only exists on JDK11+, is there mechanism to make this JDK distinction in the tests somehow?

Confusingly named, but you can stick tests that only work on JDK 11+ in the `jdk17-unit-tests` module.  E.g., see this commit:

https://github.com/uber/NullAway/commit/3c3c9f60bbb5cf1a137309bf07ade25babd70c68

XN137 commented:
@msridhar could you approve the workflows? to confirm the JDK8 problem is now fixed - thanks!

lazaroclapp reviewed (commented):
Really appreciate the thorough PR and mostly it looks good to me, with just two requests:

1. Let's refactor the new code added to `onDataflowVisitMethodInvocation` to be even further extracted into private methods and also avoid repetition.
2. If possible, let's actually check if we can _trivially_ handle `assertThat(obj.getOptionalField()).isPresent()` and `assertThat(map.get(CONST_KEY)).isPresent()`. I think we can, easily, but if it turns out to be any more involved than a couple lines and the tests, I am happy to leave it for later! 🙂

XN137 commented:
@lazaroclapp thanks for the review, i think i have addressed all your comments. let me know if something is missing.

wondering: do you guys squash on merge?

msridhar commented:
> wondering: do you guys squash on merge?

Yes, we squash on merge

lazaroclapp commented on nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java:
```diff
@@ -168,35 +170,115 @@ public boolean includeApInfoInSavedContext(AccessPath accessPath, VisitorState s
+    }
+  }
+
+  private Optional<MethodInvocationNode> getMethodForBooleanParam(MethodInvocationNode node) {
```
Wonder if we can name this something like `getJUnitMethodForBooleanParam` and provide a quick javadoc. I get that this method is private and also that it will technically work on any assertion method which takes a single boolean, _but_ I think having some clue about its usage when looking at this method in isolation, beyond the comment right below, would be useful.

XN137 replied:
i wasnt happy with the method name but i thought it should prefer to be named after what it does, not after what it is used for...
looking at it again, the method body can be simplified, so i've decided to inline this logic again in [5f95a6f](https://github.com/uber/NullAway/pull/718/commits/5f95a6f6429ecba5dfe7c0a11d5384e3eb70e410)

let me know if you want me to split it out again with the suggested name

lazaroclapp replied:
If we want to use what the method does strictly, then maybe `getFirstBooleanParamAsMethodInvocation` or something similar. That said, I agree the fix in [5f95a6f](https://github.com/uber/NullAway/pull/718/commits/5f95a6f6429ecba5dfe7c0a11d5384e3eb70e410) is cleaner! 🙂

lazaroclapp commented:
> > wondering: do you guys squash on merge?
> 
> Yes, we squash on merge

Specifically: our default practice is to squash on merge and use the PR's title and description for the commit (without the list of commit messages), sometimes editing it a bit if stuff was changed during review 🙂 When practical, the PR title also becomes a line in our CHANGELOG when releasing a new version.

p.s. I saw some question about access-paths on my email updates, but not currently on the PR. Gonna assume you figured it out, but happy to give more details if needed!

lazaroclapp reviewed (commented):
One tiny request for a method rename and clarifying doc on a single private method, otherwise this LGTM! @msridhar?

msridhar commented:
This LGTM as well!

lazaroclapp commented:
> This LGTM as well!

Cool, we can approve and land as soon as the name change and comment are there for `getMethodForBooleanParam`. Thanks again for the thorough contribution @XN137 !

XN137 commented:
should be ready for review again, sorry i prefer rebasing over merge commits...
but only the last review commit is new since you last looked.

in case this gets merged:
please make sure to use the same email from my commits - not my primary account one (let me know if that's not possible for some reason), thanks

msridhar commented:
> in case this gets merged:
> 
> please make sure to use the same email from my commits - not my primary account one (let me know if that's not possible for some reason), thanks

Do you know what will happen if we squash merge via GitHub UI? Not sure if the desired email address will be used

lazaroclapp commented:
> > in case this gets merged:
> > please make sure to use the same email from my commits - not my primary account one (let me know if that's not possible for some reason), thanks
> 
> Do you know what will happen if we squash merge via GitHub UI? Not sure if the desired email address will be used

The email that will be used as far as GitHub is concerned is as follows: `This commit will be authored by xn137@gmx.de`, @XN137 can you confirm if this is what you want or not before we approve and squash? (I am 100% onboard with this PR landing as is, but want to check this before I approve it in the UI)

XN137 commented:
> The email that will be used as far as GitHub is concerned is as follows: ...

yes that's the right email and matches up with the one used for these commits.
Sorry for the overhead but i felt the need to mention this, because it seems that on the last PR another email was used for the merge (i got multiples in my github account).

lazaroclapp commented:
> yes that's the right email and matches up with the one used for these commits. 

Cool! Will merge in a second.

It is actually a bit hard to get the email used for individual commits from the GitHub UI (it just shows your GitHub user profile instead in all places I saw). Trivial to find out by actually fetching this locally on `git` proper, but I was a bit lazy there, sorry 🙂

lazaroclapp reviewed (approved):
LGTM! Once again, thanks for the contribution(s)!

msridhar commented:
> Sorry for the overhead but i felt the need to mention this, because it seems that on the last PR another email was used for the merge (i got multiples in my github account).

I just did a squash & merge from the UI last time, so not sure why it picked the wrong email address...

Thanks again for the contribution!

XN137 commented:
thank you for the productive review and quick turn around times across the board!

wondering: is the release cadence described somewhere? :sweat_smile:
i know it will be more problematic due to the other "breaking change" but having both will be useful for introducing nullaway to our codebase.

lazaroclapp commented:
> wondering: is the release cadence described somewhere? 😅 i know it will be more problematic due to the other "breaking change" but having both will be useful for introducing nullaway to our codebase.

No formal release cadence. But just to avoid keeping you in the dark here: I was running some internal tests before releasing, trying to get a version out yesterday/today, while I ran into a two unrelated bugs uncovered by this release. They shouldn't be big changes, but might hold the release a few days 😓
